### DISCUSSION TITLE
Ensuring adherence to existing TypeScript interface

### DISCUSSION THREAD

**Discussion by sdc395** [sinclairzx81/typebox#449] [Q&A] [answered]
Hi all

I've tried to find the answer to this question myself, but no luck.  Is there a utility type in TypeBox that will let me check a schema's adherence to an exiting type.  Something like Superstruct's `Describe` type.  To illustrate...

```
interface User {
    name: string;
    age: number;
}

const user: Describe<User> = Type.Object({
    name: Type.String(),
    age: Type.Number(),
});
```

Thanks

**Comment by sinclairzx81** [comment, marked as answer]
@sdc395 Hi,

Sorry for the delay in reply on this one (I must have missed this question). TypeBox doesn't have a `Describe<T>` utility type built in, but it does have types to express each schematic. By convention, TypeBox types are prefixed with `T`, so you can enforce structures in the following way.

```typescript
import { Type, TObject, TString, TNumber } from '@sinclair/typebox'

const user: TObject<{
   name: TString,
   age: TNumber
}> = Type.Object({
    name: Type.String(),
    age: Type.Number(),
})
```
Additionally, it's possible to implement `Describe<T>` through conditional mapping. The following is one way you can achieve this.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQPICMBWBTAxjAGhQGUYpgA7AcyOQDkBXEdbKWgIQggBtsBDCrQCqFANYUIAdwpwAvnABmUCCDgByAAIBnSrm59gUAPQwAnmGzoIADzUBYAFCPs1yLDhmLcACLYtuchYAHmQAPjgAXjhHOBQ4FxhsCgATLTgAJTxoZKCBUyIGMQlpcIB+FAwcfCCkAG0AaThKOFFsUwgFFABdAC4fPwDgYOQGrvDZcJ6YuISk1LgtMkoqOHLkUnJqOCmHWOR460SUtIomFihVlEZmVm3p-dnjuCsefhk1zleBO92UEXEpBRHI5PNg-lpblFfP5AtgatMKHwQNg+otNlRpnwqCi4KcblBHBNgQ4gA)

```typescript
import { TObject, TString, TNumber, TBoolean, TUnknown } from '@sinclair/typebox'

export type Describe<T> = 
  T extends Record<any, unknown> ? TObject<{ [K in keyof T]: Describe<T[K]> }> :
  T extends string ? TString :
  T extends number ? TNumber :
  T extends boolean ? TBoolean :
  TUnknown

type TUser = Describe<{ // hover TUser
  name: string
  age: number
}>
```
Conditional mapping arrays, tuples, unions, intersections and other constructs is an exercise ill leave for you.

Hope this helps!
S
